middleware/instrument.go

```diff
@@ -28,13 +29,29 @@ type RouteMatcher interface {
 	Match(*http.Request, *mux.RouteMatch) bool
 }
 
+// PerTenantCallback is a function that returns a tenant ID for a given request. When the returned tenant ID is not empty, it is used to label the duration histogram.
+type PerTenantCallback func(context.Context) string
+
+func (f PerTenantCallback) shouldInstrument(ctx context.Context) (tenantID string, ok bool) {
+	if f == nil {
+		return "", false
+	}
+	tenantID = f(ctx)
+	if tenantID == "" {
+		return "", false
+	}
+	return tenantID, true
+}
```

Comment:
nit: It doesn't seem like named return values are common in dskit and it doesn't appear that the serve to increase readability/ergonomics of this function. Maybe they can be dropped here?

```suggestion
func (f PerTenantCallback) shouldInstrument(ctx context.Context) (string, bool) {
	if f == nil {
		return "", false
	}
	tenantID := f(ctx)
	if tenantID == "" {
		return "", false
	}
	return tenantID, true
}
```